Lua ports of biosnoop and stacksnoop

markdrayton opened the pull request:
I ported a couple of tools while testing out the Lua stuff.

What should we do about tools now that we have Python and Lua wrappers? For me Lua is easier to manage in production as bcc-lua is completely static but it might get confusing if we have two potentially unsynchronized implementations of each tool.

4ast commented:
looks great to me.
I think in this case biosnoop.lua and stacksnoop.lua are fully equivalent to their python versions, so probably should live in tools/ directory. Then hopefully whoever modifies foo.py will not forget to do the same in foo.lua and may be eventually we'll just get rid of .py versions of biosnoop and stacksnoop if bcc-lua stands up to its potential.

mcaleavya commented:
Quick? should the original copyrights / updates not be included in the lua version as the c code is replicated? And looks like I need to start learning lua now if that's the way forward for the tools.

vmg commented on examples/lua/biosnoop.lua:
```diff
@@ -0,0 +1,182 @@
+
+  local function print_event(cpu, event)
+    local val = -1
+    local event_pid = tonumber(event.pid)
```
You don't really need to call `tonumber` on any of these variables. The JIT will automatically convert them to native Lua numbers. `tonumber` is only useful for `uint64_t` data because it cannot be trivially converted to a Lua integer, so calling it will force the loss of precision.

vmg commented:
Looks like a beautiful port. Thank you!

markdrayton commented:
I moved the tools to tools/, fixed the superfluous `tonumber()`, and wrapped the lines at 80 cols.

What's the etiquette for updating a PR, BTW? In this case it doesn't seem helpful to have two commits so I just squashed locally and force pushed the branch again.

markdrayton commented:
Removed the unneeded `ffi` import from `stacksnoop.lua`. This is all good to go now.

4ast commented:
squash and force update is the right call. Thanks
